lib/src/source/hosted.dart

```diff
@@ -1100,3 +1128,82 @@ class _RefAndCache {
   @override
   bool operator ==(Object other) => other is _RefAndCache && other.ref == ref;
 }
+
+const checksumHeaderName = 'x-goog-hash';
+
+/// Adds a checksum validation "tap" to the response stream and returns a
+/// wrapped `Stream` object, which should be used to consume the incoming data.
+///
+/// As chunks are received, a CRC32C checksum is updated.
+/// Once the download is completed, the final checksum is compared with
+/// the one present in the checksum response header.
+///
+/// Throws [PackageIntegrityException] if there is a checksum mismatch.
+Stream<List<int>> _validateStream(
+    Stream<List<int>> stream, int expectedChecksum, String fileName) async* {
+  final crc32c = Crc32c();
+
+  await for (final chunk in stream) {
+    crc32c.update(chunk);
+    yield chunk;
+  }
+
+  final actualChecksum = crc32c.finalize();
+
+  log.fine(
+      'Computed checksum $actualChecksum for "$fileName" with expected CRC32C '
+      'of $expectedChecksum.');
+
+  if (actualChecksum != expectedChecksum) {
+    throw PackageIntegrityException(
+      'Package archive "$fileName" has a CRC32C checksum mismatch',
```

Comment:
Can we change this to show the URL instead of the filename? The filename is an artifact of how we process the archive data.

Reply:
Can we write something like:
> Package archive downloaded from "$url" has "x-goog-hash: crc32c=$crc" which doesn't match the checksum of the archive downloaded.